### ISSUE TITLE
Prefix `await` is cumbersome to work with.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#25]
In an asynchronous Dart function, the `await expr` expression allows blocking and waiting for a future result of `expr` to complete.
This is highly convenient compared to using `Future.then`, but grammatically it's still cumbersome because await is a prefix operator with lower precedence than selectors.
Example:
```dart
var x = await (await foo.bar()).baz();
```
If you need to await an intermediate result of a longer computation chain, you need to add parentheses and go back and write the await far distanced from the operation that created the future.

If you have a cascade like:
```dart
expr
  ..bar()
  ..baz()
  ..qux();
```
and `baz` is (or becomes) asynchronous and you want to await it before continuing, then you have to rewrite it to something like:
```
var tmp = expr..bar();
await tmp.baz();
tmp..qux();  // or one dot, doesn't matter.
```

A possible improvement would be to allow `await` as a suffix operator instead of a prefix operator:
```dart
var x = foo.bar() await.baz() await;
expr
  ..bar() 
  ..baz() await
  ..qux();
```
or maybe with a dot in front:
```dart
var x = foo.bar().await.baz().await;
expr
  ..bar()
  ..baz().await
  ..qux();
```
This is still only allowed inside asynchronous functions where `await` is a reserved word, so it would not be ambiguous.
Other alternatives could be a suffix *operator* (like `!`, although that's probably high-priced syntactic real-estate, but if it is a general heuristic coercion operator, it could coerce from `Future<T>` to `T` too) or something else that allows binding `await` better in compound expressions.

**Comment by pschiffmann** [comment]
Idea: Special-case the `await` keyword to be a valid RHS for the `|>` operator from #40. It could look like this:

```Dart
/// Changes the contents of [file] to uppercase.
/// Completes when the file is written back to disk.
Future<void> contentsToUpperCase(File file) async =>
    file.readAsString()
      |> await
      |. toUpperCase()
      |> file.writeAsString
      |> await;
```

**Comment by fkettelhoit** [comment]
Is this still being considered? It looks like this would be a very small non-breaking syntactic change, but would make working with nested futures _much_ more pleasant. For example, I'm currently working with records that form a graph in a database so that each record has a `children` getter of type `Future<List<Record>>` (these graphs can be cyclic, so I do not want to immediately load the children and their descendants). Right now this forces me to write call chains like the following:

```dart
final someChild = (await (await (await db.get(someId)).children).first.children).first;
```

With the proposed postfix await this would become much less cumbersome:

```dart
final someChild = db.get(someId).await.children.await.first.children.await.first;
```

(Of course something even shorter than `.await` would be great, but now that `!` is already used for NNBD I would personally already be very happy with `.await` or any postfix solution really.)

**Comment by mateusfccp** [comment]
@fkettelhoit 

In this cases I will simply use `.then`, as function chaining is way clearer than `await` chaining...

```dart
final someChild =
    await db.get(someId)
            .then((o) => o.children)
            .then((o) => o.first.children)
            .then((o) => o.first)

```

This would be even better if we could have something like #691.

**Comment by JSANL** [comment]
To be honest I think
```dart
expr
  ..bar() 
  ..baz() await
  ..qux();
```
and 
```dart
expr
  ..bar()
  ..baz().await
  ..qux();
```
are kinda confusing because I'm just so used to seeing `await` infront of the `Future`.
I think the `..baz().await` at least shows that the `await` does indeed wait for `baz`, in the other one I thought `await` would wait for `qux()` at first.

Maybe I'm just boring, but what about this?  
```dart
expr
  ..bar() 
  ..await baz()
  ..qux();
```  
The only thing i don't like about it is that `baz` is not directly under `bar` but I think it's still good enough to quickly scan over.

**Comment by lrhn** [comment]
Putting the `await` before the method call is something I have otherwise thought of, also for other prefix operators.

The `await foo().bar()` issue is also a problem for `!foo.bar()`. If you could put prefix operators inside a selector sequence, then you could write that as `foo().!bar()` (which is perhaps more readable as `list.!isEmpty`).

It doesn't really work as an incremental change, though. To do this, it would mean that `foo.!bar.baz` should bind to the `bar`, not the `bar.baz`, but then it's inconsistent that `!foo.bar.baz` does not bind to the `foo`. We'd have to change the meaning of existing code, making it a breaking language change. For example `foo.-` is a proposed syntax for doing an operator tear-off. so `foo.-[x]` would be ambiguous.
So, not sure that feature is so great it's worth a breaking syntax change.

**Comment by munificent** [comment]
For reference, Rust [decided to use postfix syntax for `await`](https://boats.gitlab.io/blog/post/await-decision/). I like prefix `await` for most cases, but it would be nice to support a postfix form in method chains too.

**Comment by boukeversteegh** [comment]
@JSANL wrote:

> Maybe I'm just boring, but what about this?
> 
> ```dart
> expr
>   ..bar() 
>   ..await baz()
>   ..qux();
> ```

Totally in favor of this syntax!

**Comment by lrhn** [comment]
Using a prefix operator inline in a selector chain is an enticing idea.
It generalizes, so we could allow `foo.!isEmpty` or `bar.-value` as well. The issue with that is that it might impose on our wish to do `bar.-` as the minus operator tear-off.

**Comment by tatumizer** [comment]
Just wanted to point out that the current proposal treats "await" as if it were a noun whereas, in reality, it's a verb. If we agree on it being a verb, which is a reasonable characterization, then we have to consider adding parentheses: `.await()`. Without parentheses, `.await` is a tear-off.

**Comment by NoudZandbergen** [comment]
> Maybe I'm just boring, but what about this?
> 
> ```dart
> expr
>   ..bar() 
>   ..await baz()
>   ..qux();
> ```

+1 for this solution. I would encourage making it work like this too:
> ```dart
> expr
>   .bar() 
>   .await baz()
>   .qux();
> ```

**Comment by Sufilevy** [comment]
Is this being considered? Are there any developments on the topic?

**Comment by insinfo** [comment]
@lrhn
and I really like the option with dot await  `.await`

**Comment by Maqcel** [comment]
any updates?

needed async in my `WebViewController` creation and I needed to get rid of the cascade :(

**Comment by munificent** [comment]
No updates. We've been focused on larger higher priority issues (metaprogramming, primary constructors, static member shorthands, etc.).

**Comment by munificent** [comment]
> My guess is that it doesn't really work as intended

That's correct. The extension method here doesn't actually do anything. The caller would still need to await the result.

**Comment by TekExplorer** [comment]
This would certainly be super nice to have.
`.await` as a member access is very nice for chaining - though it might look a little odd to do `<Future>[].wait.await;` but I guess we'd just have to get used to it.

I actually prefer `unawaited` as a getter since were just using it to ignore/throw away the value

`unawait()` implies we're doing something to the future (like `.ignore()`) but we're not.

I actually created that extension myself a couple times.

**Comment by Dampfwalze** [comment]
One year later, is there any activity on this feature? I just came across the same problem. I was building a cascading API and realized, I will have to implement async operations as well. But then I am forced to replace my beautiful cascading API and break it up to make room for those `await` calls. Moreover, in the past, I have encountered this problem numerous times. That's why I am desperately searching for a solution for this.

Since I also enjoy Rust a lot, I like the `.await` syntax. It takes away all the quirks of the prefix syntax and becomes equivalent to any other cascading method call. It could not be simpler than that. It only goes against the view that languages like C# and JavaScript created that where pioneering async syntax. And many people hold on to that view.

I think the implementation would not be a big problem (moreover, the issue is tagged as `small-feature`), so I assume this feature is blocked by controversial uncertainty and a lack of priority. A viable plan would be to have both, the old prefix syntax, since you cannot replace it, and implement the new `.await` suffix syntax. That way developers have options, while not being restricted by a historic convention.

One note on the formatting: I would try to keep the `.await` on the same line as the call it is associated with:

```dart
myApi
  .newCall()
  .header('a', 'b')
  .request().await
  .parseBody()
  .single;

// Instead of:

myApi
  .newCall()
  .header('a', 'b')
  .request()
  .await
  .parseBody()
  .single;

// Or for cascading operator (..):

myStream
  ..pipe(a).await
  ..pipe(b).await;
```

This avoids some of the extra complexity associated to the additional `await`s. (In any case, it is definitely less complex to having to use `.then((value) => ...)` all the time!)

**Comment by rrousselGit** [comment]
^ While I would also love an `await` suffix, you can implement a cascading API using a `Future` subclass or static extensions:

```dart
extension type MyFuture(Future<Response> _response) implements Future<Response> {
  Stream<Body> parseBody() => ...
}

abstract class MyApi {
  MyFuture request();
}

// OR:

extension on FutureOr<Response> {
  Stream<Body> parseBody() => ...
}

abstract class MyApi {
  Future<Response> request();
}
```

```dart
Response response = await myApi
  .newCall()
  .header('a', 'b')
  .request();

Body body = await myApi
  .newCall()
  .header('a', 'b')
  .request()
  .parseBody()
  .single;
```

**Comment by makri-aymen** [comment]
And it supports unawaited too, because it's useful in object initialization, like this
 ```
 @FactoryMethod()
  static SynchronizationIconCubit create() => SynchronizationIconCubit(
    authService: getIt(),
    synchronizationLocalRepository: getIt(),
    periodicSynchronizationService: getIt(),
  )..unawaited.updateSynchronizationIndicatorNumber();
```

**Comment by lrhn** [comment]
> )..unawaited.updateSynchronizationIndicatorNumber();

Hmm. The plain `unawaited` returns `void`, so that's not what you're doing here, if you can call `.updateSynhronoizationIndicatorNumber()` on it

Is the `..unawaited` returning a type that has the same members as the target, but where `Future`-returning APIs return `void` instead? 

Then just having:
```dart
extension UnawaitedFuture on Future<Object?> {
  void get unawaited {}
}
```
would be simpler, so you can do:
```dart
..updateSynchronizationIndicatorNumber().unawaited;
```
on *any* future.

> implement a cascading API using a `Future` subclass or static extensions:

Lifting your `Foo` API to the `Future<Foo>` level as extensions on `Future<Foo>` definitely works, but it's also
a lot of work for a big API. You need to duplicate _every member of every type_, and then make sure the extensions 
are imported where they're needed.
Every class and method `class Foo<T1,... Tk> { ... R foo<S1,..,Sn>(args) ... }` needs an extension
```dart
extension FutureFoo<T1,...,Tk> on Future<Foo<T1,...Tk>> {
  ...
  Future<flattenedR> foo<S1, ..., Sn>(args) => this.then((v) => v.foo<S1,...,Sn>(args));
  ...
}
```

You can use extension types, `extension type FutureBar<...>(Future<Bar<...>> _)`, and even `implements Future<Bar<...>>` as long as the `Future` API doesn't conflict with the `Bar` api. Or even then, it can choose override the `Future` API. (That's an improvement over plain extensions, if `Foo` has a `then` method, it the `Future<Foo>.then` will take precedence over the extension, with no way to get around that other than calling it fx `thenFoo` instead.)
If your own API returns the `FutureBar` type instead of `Future<Bar>`, then your clients don't even need to do the wrapping. (But you can't do that from an `async` function, #4607.)

I'd _rather_ have the language give me `.await` as a suffix than having to code-generate a copy of my entire API.
Even if that is completely doable, even automatable. And really _preferrably_ automated, otherwise it's just a duplicate that's bound to get out of sync at the most inopportune moment.
